File: 706-b-interesting-drink/drink.py
_ = int(input())

prices = list(map(int, input().split()))

prices.sort()

days_num = int(input())

# find smallest prices
def find_smallest(budget, prices, left=0, right=len(prices)):

    if left >= right:
        return left

    middle = (right + left) // 2

    # print(left, middle, right, prices[middle] > budget)

    if prices[middle] > budget:
        return find_smallest(budget, prices, left, middle)
    else:
        return find_smallest(budget, prices, middle + 1, right)


for _ in range(days_num):
    q = int(input())

    print(find_smallest(q, prices))


# print(*prices)

File: 706-b-interesting-drink/drink.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>

using namespace std;

int find_match(int budget, vector<int>::iterator begin, vector<int>::iterator end);

int main()
{
    int n;
    cin >> n;

    vector<int> prices;
    for (int i = 0; i < n; i++)
    {
        int x;
        cin >> x;
        prices.push_back(x);
    }

    sort(prices.begin(), prices.end());

    int q;
    cin >> q;

    for (int i = 0; i < q; i++)
    {
        int b;
        cin >> b;
        cout << find_match(b, prices.begin(), prices.end()) << endl;
    }
}

int find_match(int budget, vector<int>::iterator begin, vector<int>::iterator end)
{
    auto start = begin;

    while (distance(begin, end))
    {
        auto middle = begin + distance(begin, end) / 2;

        if (*middle > budget)
        {
            end = middle;
        }
        else
        {
            begin = middle + 1;
        }
    }
    return distance(start, begin);
}
